openeo/extra/job_management.py

```diff
@@ -0,0 +1,290 @@
+                                    exc_info=True,
+                                )
+                                df.loc[i, "status"] = "start_failed"
+                            else:
+                                df.loc[
+                                    i, "start_time"
+                                ] = datetime.datetime.now().isoformat()
+                                if job:
+                                    df.loc[i, "id"] = job.job_id
+                                    with ignore_connection_errors(context="get status"):
+                                        status = job.status()
+                                        df.loc[i, "status"] = status
+                                        if status == "created":
+                                            # start job if not yet done by callback
+                                            try:
+                                                job.start_job()
+                                                df.loc[i, "status"] = job.status()
+                                            except OpenEoApiError as e:
+                                                _log.error(e)
+                                                df.loc[i, "status"] = "start_failed"
+                                else:
+                                    df.loc[i, "status"] = "skipped"
+
+                            persists(df, output_file)
+
+            time.sleep(self.poll_sleep)
+
+    def on_job_done(self, job: BatchJob, row):
+        """
+        Handles jobs that have finished. Can be overridden to provide custom behaviour.
+
+        Default implementation downloads the results into a folder containing the title.
+
+        :param job: The job that has finished.
+        :param row: DataFrame row containing the job's metadata.
+        """
+        # TODO: param `row` is never accessed in this method. Remove it? Is this intended for future use?
+
+        job_metadata = job.describe_job()
+        job.get_results().download_files(target=job_metadata["title"])
+        with open(Path(job_metadata["title"]) / f"job_{job.job_id}.json", "w") as f:
+            json.dump(job_metadata, f, ensure_ascii=False)
+
+    def on_job_error(self, job: BatchJob, row):
+        """
+        Handles jobs that stopped with errors.  Can be overridden to provide custom behaviour.
+
+        Default implementation writes the error logs to a JSON file.
+
+        :param job: The job that has finished.
+        :param row: DataFrame row containing the job's metadata.
+        """
+        # TODO: param `row` is never accessed in this method. Remove it? Is this intended for future use?
+
+        logs = job.logs()
+        error_logs = [l for l in logs if l.level.lower() == "error"]
+        job_metadata = job.describe_job()
+
+        title = job_metadata["title"]
+        if len(error_logs) > 0:
+            f"job_{title}_errors.json".write_text(json.dumps(error_logs, indent=2))
```

Comment:
I don't think `str.write_text` works, I guess a `Path()` around the str is missing.

also, instead of writing to working dir, I would use a configurable root dir from `__ini__` as described above

Reply:
Indeed this produced errors. Updated along with the other stuff related to configurable root directory.
I added a basic test case to check the error logging works.